File: backend/api/projects.py
"""
Projects API

项目 CRUD 端点。
"""

from uuid import UUID
import structlog
from fastapi import APIRouter, Depends, HTTPException, status

from backend.schemas.project import ProjectCreate, ProjectResponse, ProjectUpdate
from backend.schemas.common import SuccessResponse, PaginatedResponse
from backend.services import get_db_service, DatabaseService
from backend.api.deps import get_current_user_id

router = APIRouter(prefix="/projects", tags=["Projects"])
logger = structlog.get_logger(__name__)


@router.post(
    "", response_model=SuccessResponse[ProjectResponse], status_code=status.HTTP_201_CREATED
)
async def create_project(
    data: ProjectCreate,
    user_id: str = Depends(get_current_user_id),
    db: DatabaseService = Depends(get_db_service),
):
    """创建项目"""
    project = await db.create_project(user_id, data)
    logger.info("Project created", project_id=str(project.id))
    return SuccessResponse.of(project)


@router.get("", response_model=PaginatedResponse[ProjectResponse])
async def list_projects(
    page: int = 1,
    page_size: int = 20,
    user_id: str = Depends(get_current_user_id),
    db: DatabaseService = Depends(get_db_service),
):
    """获取项目列表"""
    offset = (page - 1) * page_size
    projects = await db.list_projects(user_id, limit=page_size, offset=offset)
    total = await db.count_projects(user_id)
    return PaginatedResponse.of(projects, total, page, page_size)


@router.get("/{project_id}", response_model=SuccessResponse[ProjectResponse])
async def get_project(
    project_id: UUID,
    db: DatabaseService = Depends(get_db_service),
):
    """获取项目详情"""
    project = await db.get_project(str(project_id))
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return SuccessResponse.of(project)


@router.patch("/{project_id}", response_model=SuccessResponse[ProjectResponse])
async def update_project(
    project_id: UUID,
    data: ProjectUpdate,
    db: DatabaseService = Depends(get_db_service),
):
    """更新项目"""
    project = await db.update_project(str(project_id), data)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return SuccessResponse.of(project)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_project(
    project_id: UUID,
    db: DatabaseService = Depends(get_db_service),
):
    """删除项目"""
    success = await db.delete_project(str(project_id))
    if not success:
        raise HTTPException(status_code=404, detail="Project not found")


@router.post(
    "/temp", response_model=SuccessResponse[ProjectResponse], status_code=status.HTTP_201_CREATED
)
async def create_temp_project(
    user_id: str = Depends(get_current_user_id),
    db: DatabaseService = Depends(get_db_service),
):
    """创建临时项目"""
    # 限制临时项目数量（最多5个）
    temp_count = await db.count_temp_projects(user_id)
    if temp_count >= 5:
        # 删除最旧的临时项目
        await db.delete_oldest_temp_project(user_id)

    # 创建临时项目
    data = ProjectCreate(name="未命名项目")
    project = await db.create_temp_project(user_id, data)
    logger.info("Temp project created", project_id=str(project.id))
    return SuccessResponse.of(project)


@router.post("/{project_id}/save", response_model=SuccessResponse[ProjectResponse])
async def save_temp_project(
    project_id: UUID,
    data: ProjectUpdate,
    db: DatabaseService = Depends(get_db_service),
):
    """将临时项目转为正式项目"""
    # 验证项目是否存在
    existing = await db.get_project(str(project_id))
    if not existing:
        raise HTTPException(status_code=404, detail="Project not found")

    project = await db.save_temp_project(str(project_id), data)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    logger.info("Temp project saved", project_id=str(project_id))
    return SuccessResponse.of(project)

File: new-fronted/src/api/services/projects.ts
import { client } from '../client';
import type { components } from '@/types/api';

// 使用生成的类型
type ProjectResponse = components['schemas']['ProjectResponse'];
type ProjectCreate = components['schemas']['ProjectCreate'];
type ProjectUpdate = components['schemas']['ProjectUpdate'];

/**
 * 项目服务 - 直接使用 OpenAPI 生成类型
 * 后端已配置 camelCase alias，无需手动转换
 */
export const projectsService = {
  /**
   * 获取项目列表
   */
  async listProjects(page = 1, pageSize = 20) {
    const { data, error } = await client.GET('/api/projects', {
      params: {
        query: { page, page_size: pageSize },
      },
    });

    if (error) throw error;
    // 直接返回 data，它是 PaginatedResponse 结构，包含 data (数组) 和分页信息
    return data!;
  },

  /**
   * 获取单个项目
   */
  async getProject(projectId: string) {
    const { data, error } = await client.GET('/api/projects/{project_id}', {
      params: {
        path: { project_id: projectId },
      },
    });

    if (error) throw error;
    return data!;
  },

  /**
   * 创建项目
   */
  async createProject(project: ProjectCreate) {
    console.log('[createProject] 发送请求体:', project);
    console.log('[createProject] 请求体类型:', typeof project);

    const { data, error } = await client.POST('/api/projects', {
      body: project,
    });

    console.log('[createProject] 响应:', { data, error });

    if (error) throw error;
    return data!;
  },

  /**
   * 更新项目
   */
  async updateProject(projectId: string, updates: ProjectUpdate) {
    const { data, error } = await client.PATCH('/api/projects/{project_id}', {
      params: {
        path: { project_id: projectId },
      },
      body: updates,
    });

    if (error) throw error;
    return data!;
  },

  /**
   * 删除项目
   */
  async deleteProject(projectId: string) {
    const { error } = await client.DELETE('/api/projects/{project_id}', {
      params: {
        path: { project_id: projectId },
      },
    });

    if (error) throw error;
  },
  /**
   * 创建临时项目
   * 调用后端原生接口，支持自动清理
   */
  async createTempProject() {
    const { data, error } = await client.POST('/api/projects/temp');
    if (error) throw error;
    return data!;
  },

  /**
   * 将临时项目转正
   */
  async saveTempProject(projectId: string, updates: ProjectUpdate) {
    const { data, error } = await client.POST('/api/projects/{project_id}/save', {
      params: {
        path: { project_id: projectId },
      },
      body: updates,
    });

    if (error) throw error;
    return data!;
  },
};

export type { ProjectResponse, ProjectCreate, ProjectUpdate };
